File: lib/widgets/menu_card.dart
import 'package:flutter/material.dart';
import '../models/cart_item_model.dart';
import '../models/menu_model.dart';
import '../repositories/restaurant_repository.dart';
import '../theme/app_colors.dart';

class MenuCard extends StatelessWidget {
  final MenuModel menu;
  final CartItemModel? cartItem;
  final VoidCallback onAddToCart;
  final ValueChanged<int> onQuantityChange;
  final ValueChanged<CartItemModel> onOpenNotes;

  const MenuCard({
    super.key,
    required this.menu,
    required this.cartItem,
    required this.onAddToCart,
    required this.onQuantityChange,
    required this.onOpenNotes,
  });

  IconData _getCategoryIcon(String categoryName) {
    final lower = categoryName.toLowerCase();
    if (lower.contains('minum') || lower.contains('teh') || lower.contains('kopi')) {
      return Icons.local_bar;
    }
    if (lower.contains('snack') || lower.contains('camilan') || lower.contains('pisang')) {
      return Icons.bakery_dining;
    }
    if (lower.contains('ikan') || lower.contains('laut') || lower.contains('seafood')) {
      return Icons.set_meal;
    }
    if (lower.contains('nasi') || lower.contains('ayam') || lower.contains('bebek') || lower.contains('goreng')) {
      return Icons.lunch_dining;
    }
    return Icons.restaurant;
  }

  @override
  Widget build(BuildContext context) {
    final isOutOfStock = menu.stock <= 0;
    final inCartQty = cartItem?.quantity ?? 0;

    return Opacity(
      opacity: isOutOfStock ? 0.6 : 1.0,
      child: Container(
        decoration: BoxDecoration(
          color: Colors.white,
          borderRadius: BorderRadius.circular(20),
          border: Border.all(
            color: inCartQty > 0
                ? AppColors.primary
                : AppColors.surfaceLightBorder.withValues(alpha: 0.5),
            width: inCartQty > 0 ? 1.5 : 1.0,
          ),
          boxShadow: [
            BoxShadow(
              color: Colors.black.withValues(alpha: 0.04),
              blurRadius: 8,
              offset: const Offset(0, 2),
            ),
          ],
        ),
        padding: const EdgeInsets.all(12),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            // Top: Category Icon + Stock Badge
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Container(
                  width: 38,
                  height: 38,
                  decoration: BoxDecoration(
                    color: AppColors.primaryContainer,
                    borderRadius: BorderRadius.circular(12),
                  ),
                  child: Icon(
                    _getCategoryIcon(menu.categoryName),
                    color: AppColors.onPrimaryContainer,
                    size: 20,
                  ),
                ),
                Container(
                  padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 3),
                  decoration: BoxDecoration(
                    color: isOutOfStock
                        ? AppColors.error.withValues(alpha: 0.12)
                        : AppColors.success.withValues(alpha: 0.12),
                    borderRadius: BorderRadius.circular(20),
                  ),
                  child: Text(
                    isOutOfStock ? 'Habis' : 'Stok: ${menu.stock}',
                    style: TextStyle(
                      fontSize: 10,
                      fontWeight: FontWeight.bold,
                      color: isOutOfStock ? AppColors.error : AppColors.success,
                    ),
                  ),
                ),
              ],
            ),

            const SizedBox(height: 8),

            // Middle: Name & Description
            Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  menu.name,
                  style: const TextStyle(
                    fontSize: 14,
                    fontWeight: FontWeight.bold,
                    color: AppColors.textPrimaryLight,
                  ),
                  maxLines: 2,
                  overflow: TextOverflow.ellipsis,
                ),
                if (menu.description.isNotEmpty) ...[
                  const SizedBox(height: 2),
                  Text(
                    menu.description,
                    style: const TextStyle(
                      fontSize: 11,
                      color: AppColors.textSecondaryLight,
                      height: 1.2,
                    ),
                    maxLines: 2,
                    overflow: TextOverflow.ellipsis,
                  ),
                ],
              ],
            ),

            const SizedBox(height: 8),

            // Note preview if exists
            if (cartItem != null && cartItem!.notes.isNotEmpty) ...[
              Container(
                width: double.infinity,
                padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 3),
                decoration: BoxDecoration(
                  color: AppColors.primaryContainer.withValues(alpha: 0.5),
                  borderRadius: BorderRadius.circular(6),
                ),
                child: Text(
                  '📝 ${cartItem!.notes}',
                  style: const TextStyle(
                    fontSize: 10,
                    fontWeight: FontWeight.w600,
                    color: AppColors.onPrimaryContainer,
                  ),
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                ),
              ),
              const SizedBox(height: 6),
            ],

            // Bottom: Price & Action Controls
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Text(
                  RestaurantRepository.formatRupiah(menu.price),
                  style: const TextStyle(
                    fontSize: 13,
                    fontWeight: FontWeight.w800,
                    color: AppColors.primary,
                  ),
                ),

                if (inCartQty > 0)
                  Row(
                    mainAxisSize: MainAxisSize.min,
                    children: [
                      // Notes button
                      InkWell(
                        onTap: () => onOpenNotes(cartItem!),
                        borderRadius: BorderRadius.circular(16),
                        child: Container(
                          width: 26,
                          height: 26,
                          decoration: BoxDecoration(
                            color: cartItem!.notes.isNotEmpty
                                ? AppColors.primary.withValues(alpha: 0.15)
                                : AppColors.surfaceLightChip,
                            shape: BoxShape.circle,
                          ),
                          child: Icon(
                            Icons.edit_note,
                            size: 16,
                            color: cartItem!.notes.isNotEmpty
                                ? AppColors.primary
                                : AppColors.textSecondaryLight,
                          ),
                        ),
                      ),
                      const SizedBox(width: 4),

                      // Minus button
                      InkWell(
                        onTap: () => onQuantityChange(inCartQty - 1),
                        borderRadius: BorderRadius.circular(16),
                        child: Container(
                          width: 26,
                          height: 26,
                          decoration: const BoxDecoration(
                            color: AppColors.surfaceLightChip,
                            shape: BoxShape.circle,
                          ),
                          child: const Icon(
                            Icons.remove,
                            size: 14,
                            color: AppColors.textPrimaryLight,
                          ),
                        ),
                      ),
                      Padding(
                        padding: const EdgeInsets.symmetric(horizontal: 4),
                        child: Text(
                          '$inCartQty',
                          style: const TextStyle(
                            fontSize: 13,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                      ),

                      // Plus button
                      InkWell(
                        onTap: inCartQty < menu.stock ? () => onQuantityChange(inCartQty + 1) : null,
                        borderRadius: BorderRadius.circular(16),
                        child: Container(
                          width: 26,
                          height: 26,
                          decoration: BoxDecoration(
                            color: inCartQty < menu.stock ? AppColors.primary : AppColors.surfaceLightChip,
                            shape: BoxShape.circle,
                          ),
                          child: Icon(
                            Icons.add,
                            size: 14,
                            color: inCartQty < menu.stock ? Colors.white : AppColors.textSecondaryLight,
                          ),
                        ),
                      ),
                    ],
                  )
                else
                  ElevatedButton(
                    onPressed: isOutOfStock ? null : onAddToCart,
                    style: ElevatedButton.styleFrom(
                      backgroundColor: AppColors.primary,
                      foregroundColor: Colors.white,
                      padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 4),
                      minimumSize: const Size(60, 30),
                      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(10)),
                      elevation: 0,
                    ),
                    child: const Row(
                      mainAxisSize: MainAxisSize.min,
                      children: [
                        Icon(Icons.add, size: 14),
                        SizedBox(width: 2),
                        Text('Pesan', style: TextStyle(fontSize: 11, fontWeight: FontWeight.bold)),
                      ],
                    ),
                  ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

File: app/src/main/java/com/example/ui/components/MenuCard.kt
package com.example.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.BakeryDining
import androidx.compose.material.icons.filled.Coffee
import androidx.compose.material.icons.filled.EditNote
import androidx.compose.material.icons.filled.LocalBar
import androidx.compose.material.icons.filled.LunchDining
import androidx.compose.material.icons.filled.Remove
import androidx.compose.material.icons.filled.Restaurant
import androidx.compose.material.icons.filled.SetMeal
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.FilledTonalIconButton
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.data.model.CartItem
import com.example.data.model.MenuEntity
import com.example.data.repository.RestaurantRepository
import com.example.ui.theme.ErrorRed
import com.example.ui.theme.PurpleOnPrimaryContainer
import com.example.ui.theme.PurplePrimary
import com.example.ui.theme.PurplePrimaryContainer
import com.example.ui.theme.SuccessGreen
import com.example.ui.theme.SurfaceLightBorder

@Composable
fun MenuCard(
    menu: MenuEntity,
    cartItem: CartItem?,
    onAddToCart: () -> Unit,
    onQuantityChange: (Int) -> Unit,
    onOpenNotes: (CartItem) -> Unit,
    modifier: Modifier = Modifier
) {
    val isOutOfStock = menu.stock <= 0
    val inCartQty = cartItem?.quantity ?: 0

    Card(
        modifier = modifier
            .fillMaxWidth()
            .testTag("menu_card_${menu.id}")
            .alpha(if (isOutOfStock) 0.6f else 1.0f),
        shape = RoundedCornerShape(20.dp),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surface
        ),
        elevation = CardDefaults.cardElevation(defaultElevation = 1.dp),
        border = if (inCartQty > 0) androidx.compose.foundation.BorderStroke(1.5.dp, PurplePrimary)
        else androidx.compose.foundation.BorderStroke(1.dp, SurfaceLightBorder.copy(alpha = 0.5f))
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(14.dp)
        ) {
            // Icon thumbnail + stock badge
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Box(
                    modifier = Modifier
                        .size(42.dp)
                        .clip(RoundedCornerShape(12.dp))
                        .background(PurplePrimaryContainer),
                    contentAlignment = Alignment.Center
                ) {
                    Icon(
                        imageVector = getMenuIcon(menu.categoryName),
                        contentDescription = menu.categoryName,
                        tint = PurpleOnPrimaryContainer,
                        modifier = Modifier.size(22.dp)
                    )
                }

                Surface(
                    shape = RoundedCornerShape(20.dp),
                    color = if (isOutOfStock) ErrorRed.copy(alpha = 0.12f) else SuccessGreen.copy(alpha = 0.12f)
                ) {
                    Text(
                        text = if (isOutOfStock) "Habis" else "Stok: ${menu.stock}",
                        fontSize = 11.sp,
                        fontWeight = FontWeight.Bold,
                        color = if (isOutOfStock) ErrorRed else SuccessGreen,
                        modifier = Modifier.padding(horizontal = 8.dp, vertical = 3.dp)
                    )
                }
            }

            Spacer(modifier = Modifier.height(10.dp))

            // Menu title
            Text(
                text = menu.name,
                fontSize = 15.sp,
                fontWeight = FontWeight.Bold,
                color = MaterialTheme.colorScheme.onSurface,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis
            )

            if (menu.description.isNotBlank()) {
                Spacer(modifier = Modifier.height(3.dp))
                Text(
                    text = menu.description,
                    fontSize = 12.sp,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                    lineHeight = 16.sp
                )
            }

            Spacer(modifier = Modifier.height(12.dp))

            // Price and Action controls
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = RestaurantRepository.formatRupiah(menu.price),
                    fontSize = 15.sp,
                    fontWeight = FontWeight.ExtraBold,
                    color = PurplePrimary
                )

                if (inCartQty > 0) {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(4.dp)
                    ) {
                        // Notes button
                        Box(
                            modifier = Modifier
                                .size(30.dp)
                                .clip(CircleShape)
                                .background(
                                    if (!cartItem?.notes.isNullOrBlank()) PurplePrimary.copy(alpha = 0.2f)
                                    else MaterialTheme.colorScheme.surfaceVariant
                                )
                                .clickable { cartItem?.let { onOpenNotes(it) } },
                            contentAlignment = Alignment.Center
                        ) {
                            Icon(
                                imageVector = Icons.Default.EditNote,
                                contentDescription = "Catatan",
                                tint = if (!cartItem?.notes.isNullOrBlank()) PurplePrimary else MaterialTheme.colorScheme.onSurfaceVariant,
                                modifier = Modifier.size(16.dp)
                            )
                        }

                        // Minus
                        FilledTonalIconButton(
                            onClick = { onQuantityChange(inCartQty - 1) },
                            modifier = Modifier.size(30.dp),
                            shape = CircleShape,
                            colors = IconButtonDefaults.filledTonalIconButtonColors(
                                containerColor = MaterialTheme.colorScheme.surfaceVariant
                            )
                        ) {
                            Icon(
                                imageVector = Icons.Default.Remove,
                                contentDescription = "Kurang",
                                modifier = Modifier.size(14.dp)
                            )
                        }

                        Text(
                            text = inCartQty.toString(),
                            fontSize = 14.sp,
                            fontWeight = FontWeight.Bold,
                            modifier = Modifier.padding(horizontal = 4.dp)
                        )

                        // Plus
                        FilledTonalIconButton(
                            onClick = { onQuantityChange(inCartQty + 1) },
                            modifier = Modifier.size(30.dp),
                            shape = CircleShape,
                            enabled = inCartQty < menu.stock,
                            colors = IconButtonDefaults.filledTonalIconButtonColors(
                                containerColor = PurplePrimary,
                                contentColor = Color.White
                            )
                        ) {
                            Icon(
                                imageVector = Icons.Default.Add,
                                contentDescription = "Tambah",
                                modifier = Modifier.size(14.dp)
                            )
                        }
                    }
                } else {
                    Surface(
                        onClick = onAddToCart,
                        enabled = !isOutOfStock,
                        shape = RoundedCornerShape(12.dp),
                        color = if (isOutOfStock) MaterialTheme.colorScheme.surfaceVariant else PurplePrimary,
                        modifier = Modifier.testTag("add_button_${menu.id}")
                    ) {
                        Row(
                            modifier = Modifier.padding(horizontal = 12.dp, vertical = 6.dp),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Icon(
                                imageVector = Icons.Default.Add,
                                contentDescription = null,
                                tint = if (isOutOfStock) MaterialTheme.colorScheme.onSurfaceVariant else Color.White,
                                modifier = Modifier.size(15.dp)
                            )
                            Spacer(modifier = Modifier.width(4.dp))
                            Text(
                                text = "Pesan",
                                fontSize = 12.sp,
                                fontWeight = FontWeight.Bold,
                                color = if (isOutOfStock) MaterialTheme.colorScheme.onSurfaceVariant else Color.White
                            )
                        }
                    }
                }
            }

            // Show active note indicator if any
            if (!cartItem?.notes.isNullOrBlank()) {
                Spacer(modifier = Modifier.height(6.dp))
                Surface(
                    shape = RoundedCornerShape(8.dp),
                    color = PurplePrimaryContainer.copy(alpha = 0.5f),
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Text(
                        text = "📝 ${cartItem?.notes}",
                        fontSize = 11.sp,
                        color = PurpleOnPrimaryContainer,
                        fontWeight = FontWeight.Medium,
                        modifier = Modifier.padding(horizontal = 8.dp, vertical = 3.dp),
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis
                    )
                }
            }
        }
    }
}

private fun getMenuIcon(categoryName: String): ImageVector {
    return when {
        categoryName.contains("Minum", ignoreCase = true) || categoryName.contains("Teh", ignoreCase = true) || categoryName.contains("Kopi", ignoreCase = true) -> Icons.Default.LocalBar
        categoryName.contains("Snack", ignoreCase = true) || categoryName.contains("Camilan", ignoreCase = true) || categoryName.contains("Pisang", ignoreCase = true) -> Icons.Default.BakeryDining
        categoryName.contains("Ikan", ignoreCase = true) || categoryName.contains("Laut", ignoreCase = true) -> Icons.Default.SetMeal
        categoryName.contains("Nasi", ignoreCase = true) || categoryName.contains("Ayam", ignoreCase = true) || categoryName.contains("Goreng", ignoreCase = true) -> Icons.Default.LunchDining
        else -> Icons.Default.Restaurant
    }
}
